src/freenet/client/async/KeyListenerTracker.java

```diff
@@ -95,61 +102,197 @@ protected static int fixRetryCount(int retryCount) {
+							singleKeyListeners.put(wrapper,newListeners);
+						}
+					}
+				}
+			} else {
+				ret = keyListeners.remove(listener);
+			}
 			listener.onRemove();
 		}
 		if (logMINOR)
-			Logger.minor(this, "Removed pending keys from "+this+" : size now "+keyListeners.size()+" : "+listener, new Exception("debug"));
+			Logger.minor(this, "Removed pending keys from "+this+" : size now "+this.keyListeners.size()+"/"+singleKeyListeners.size()+" : "+listener, new Exception("debug"));
 		return ret;
 	}
 	
-	public synchronized boolean removePendingKeys(HasKeyListener hasListener) {
-		boolean found = false;
-		for(Iterator<KeyListener> i = keyListeners.iterator();i.hasNext();) {
-			KeyListener listener = i.next();
-			if(listener == null) {
-				i.remove();
-				Logger.error(this, "Null KeyListener in removePendingKeys()");
-				continue;
+	public boolean removePendingKeys(HasKeyListener hasListener) {
+		boolean ret = false;
+		byte[] wantedKey = hasListener.getWantedKey();
+		ByteArrayWrapper wrapper = wantedKey != null ? new ByteArrayWrapper(saltKey(wantedKey)) : null;
+		synchronized(this) {
+			if(wantedKey != null) {
+				Object o = singleKeyListeners.get(wrapper);
+				if(o == null) {
+					// do nothing
+				} else if(o instanceof KeyListener) {
+					if((ret = ((KeyListener)o).getHasKeyListener() == hasListener)) {
+						singleKeyListeners.remove(wrapper);
+						((KeyListener)o).onRemove();
+					}
+				} else {
+					@SuppressWarnings("unchecked")
+					KeyListener[] listeners = (KeyListener[])o;
+					KeyListener[] newListeners = new KeyListener[listeners.length-1];
+					int x = 0;
+					String msg = logMINOR ? "" : null;
+					for (KeyListener l : listeners) {
+						if(l.getHasKeyListener() == hasListener) {
+							ret = true;
+							l.onRemove();
+							if (logMINOR)
+								msg = msg + " : " + l;
+							continue;
+						}
+						if(x == newListeners.length) {
+							assert(!ret);
+							break;
+						}
+						newListeners[x++] = l;
+					}
+					if(!ret) {
+						// do nothing
+					} else {
+						if(x < newListeners.length)
+							newListeners = Arrays.copyOf(newListeners, x);
+						if(newListeners.length == 0) {
+							singleKeyListeners.remove(wrapper);
+						} else if(newListeners.length == 1) {
+							singleKeyListeners.put(wrapper,newListeners[0]);
+						} else {
+							singleKeyListeners.put(wrapper,newListeners);
+						}
+						if (logMINOR)
+							Logger.minor(this, "Removed pending keys from " + this + " : size now "+this.keyListeners.size()+"/"+singleKeyListeners.size() + msg);
+					}
+				}
+				return ret;
 			}
-			if(listener.getHasKeyListener() == hasListener) {
-				found = true;
-				i.remove();
-				listener.onRemove();
-				Logger.normal(this, "Removed pending keys from "+this+" : size now "+keyListeners.size()+" : "+listener);
+			for(Iterator<KeyListener> i = keyListeners.iterator();i.hasNext();) {
+				KeyListener listener = i.next();
+				if(listener.getHasKeyListener() == hasListener) {
```

Comment:
`.equals()` should be used here instead of comparing `KeyListener`s by identity.